Memory leak when using tokio::runtime::Runtime::new() with Tokio 1.x

**Version**

`cargo tree | grep tokio`
```
│   └── tokio v1.1.0
│       └── tokio-macros v1.0.0 (proc-macro)
│   │   ├── tokio v1.1.0 (*)
│   │   └── tokio-util v0.2.0
│   │       └── tokio v0.2.22
│   └── tokio v1.1.0 (*)
└── tokio v1.1.0 (*)
```

**Platform**
`Linux 5.10.9-201.fc33.x86_64 #1 SMP Wed Jan 20 16:56:23 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux`

Fedora 33, also happens on Ubuntu 20.04

**Description**

When using Tokio rt-multi-thread, I've found a memory leak using valgrind. This is affecting Nispor project and the report is here:

```
	valgrind --trace-children=yes --leak-check=full \
	--error-exitcode=1 \
	./nispor_test 1>/dev/null
==136624== Memcheck, a memory error detector
==136624== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==136624== Using Valgrind-3.16.1 and LibVEX; rerun with -h for copyright info
==136624== Command: ./nispor_test
==136624== 
Unhandled BRIDGE_PORT_INFO 35 [0]
Unhandled BRIDGE_PORT_INFO 36 [0]
==136624== 
==136624== HEAP SUMMARY:
==136624==     in use at exit: 115,104 bytes in 279 blocks
==136624==   total heap usage: 2,508 allocs, 2,229 frees, 704,318 bytes allocated
==136624== 
==136624== 115,104 (312 direct, 114,792 indirect) bytes in 3 blocks are definitely lost in loss record 78 of 78
==136624==    at 0x4839809: malloc (vg_replace_malloc.c:307)
==136624==    by 0x4A4CDBB: alloc::alloc::alloc (alloc.rs:84)
==136624==    by 0x4A4CE79: alloc::alloc::Global::alloc_impl (alloc.rs:164)
==136624==    by 0x4A4ED99: <alloc::alloc::Global as core::alloc::AllocRef>::alloc (alloc.rs:224)
==136624==    by 0x4A4CD1C: alloc::alloc::exchange_malloc (alloc.rs:314)
==136624==    by 0x4A1681D: alloc::sync::Arc<T>::new (sync.rs:330)
==136624==    by 0x4A1A4B9: <alloc::sync::Arc<T> as core::default::Default>::default (sync.rs:2164)
==136624==    by 0x4A1576F: core::array::<impl core::default::Default for [T; _]>::default (mod.rs:365)
==136624==    by 0x4A3F04A: tokio::util::slab::Slab<T>::new (slab.rs:181)
==136624==    by 0x4A67AD9: tokio::io::driver::Driver::new (mod.rs:119)
==136624==    by 0x4A257A1: tokio::runtime::driver::create_io_stack (driver.rs:22)
==136624==    by 0x4A24B3D: tokio::runtime::driver::Driver::new (driver.rs:169)
==136624== 
==136624== LEAK SUMMARY:
==136624==    definitely lost: 312 bytes in 3 blocks
==136624==    indirectly lost: 114,792 bytes in 276 blocks
==136624==      possibly lost: 0 bytes in 0 blocks
==136624==    still reachable: 0 bytes in 0 blocks
==136624==         suppressed: 0 bytes in 0 blocks
==136624== 
==136624== For lists of detected and suppressed errors, rerun with: -s
==136624== ERROR SUMMARY: 1 errors from 1 contexts (suppressed: 0 from 0)
make: *** [Makefile:36: check] Error 1
make: Leaving directory '/home/ffmancera/pifostio/nispor/test/clib'
```

@Darksonn helped me to get a reproducer, thanks! Reproducer:

```rust
use rtnetlink::new_connection;

fn main() {
    let rt = tokio::runtime::Runtime::new().unwrap();
    rt.block_on(async {
        let (connection, handle, _) = new_connection().unwrap();
        tokio::spawn(connection);
    });
    drop(rt);

    std::thread::sleep_ms(1000);
}
```

Please, notice that it is **NOT** happening when doing:
```rust
use rtnetlink::new_connection;

fn main() {
    let rt = tokio::runtime::Runtime::new().unwrap();
    rt.block_on(async {
        // ... nothing ...
    });
    drop(rt);

    std::thread::sleep_ms(1000);
}
```

Thanks for all the help!

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -28,6 +28,7 @@ jobs:
       - clippy
       - docs
       - loom
+      - valgrind
     steps:
       - run: exit 0
 
@@ -66,6 +67,27 @@ jobs:
       - name: test tests-build --each-feature
         run: cargo hack test --each-feature
         working-directory: tests-build
+
+  valgrind:
+    name: valgrind
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
+      - name: Install Rust
+        run: rustup update stable
+
+      - name: Install Valgrind
+        run: |
+          sudo apt-get update -y
+          sudo apt-get install -y valgrind
+
+      # Compile tests
+      - name: cargo build
+        run: cargo build -p tests-integration --bin test-mem
+
+      # Run with valgrind
+      - name: Run valgrind
+        run: valgrind  --leak-check=full --show-leak-kinds=all ./target/debug/test-mem
 
   test-unstable:
     name: test tokio full --unstable

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -83,7 +83,7 @@ jobs:
 
       # Compile tests
       - name: cargo build
-        run: cargo build -p tests-integration --bin test-mem
+        run: cargo build -p tests-integration --features rt-net --bin test-mem
 
       # Run with valgrind
       - name: Run valgrind

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -83,7 +83,8 @@ jobs:
 
       # Compile tests
       - name: cargo build
-        run: cargo build -p tests-integration --features rt-net --bin test-mem
+        run: cargo build --features rt-net --bin test-mem
+        working-directory: tests-integration
 
       # Run with valgrind
       - name: Run valgrind

--- a/tests-integration/Cargo.toml
+++ b/tests-integration/Cargo.toml
@@ -5,7 +5,17 @@ authors = ["Tokio Contributors <team@tokio.rs>"]
 edition = "2018"
 publish = false
 
+[[bin]]
+name = "test-cat"
+
+[[bin]]
+name = "test-mem"
+required-features = ["rt-net"]
+
 [features]
+# For mem check
+rt-net = ["tokio/rt", "tokio/rt-multi-thread", "tokio/net"]
+
 full = [
     "macros",
     "rt",
@@ -23,6 +33,4 @@ rt-multi-thread = ["rt", "tokio/rt-multi-thread"]
 tokio = { path = "../tokio" }
 tokio-test = { path = "../tokio-test", optional = true }
 doc-comment = "0.3.1"
-
-[dev-dependencies]
 futures = { version = "0.3.0", features = ["async-await"] }

--- a/tests-integration/src/bin/test-mem.rs
+++ b/tests-integration/src/bin/test-mem.rs
@@ -8,9 +8,9 @@ fn main() {
         .unwrap();
 
     rt.block_on(async {
-        let  listener = tokio::net::TcpListener::bind("0.0.0.0:0").await.unwrap();
+        let listener = tokio::net::TcpListener::bind("0.0.0.0:0").await.unwrap();
         tokio::spawn(async move {
-            loop { 
+            loop {
                 poll_fn(|cx| listener.poll_accept(cx)).await.unwrap();
             }
         });

--- a/tests-integration/src/bin/test-mem.rs
+++ b/tests-integration/src/bin/test-mem.rs
@@ -1,0 +1,21 @@
+use futures::future::poll_fn;
+
+fn main() {
+    let rt = tokio::runtime::Builder::new_multi_thread()
+        .worker_threads(1)
+        .enable_io()
+        .build()
+        .unwrap();
+
+    rt.block_on(async {
+        let  listener = tokio::net::TcpListener::bind("0.0.0.0:0").await.unwrap();
+        tokio::spawn(async move {
+            loop { 
+                poll_fn(|cx| listener.poll_accept(cx)).await.unwrap();
+            }
+        });
+    });
+
+    std::thread::sleep(std::time::Duration::from_millis(50));
+    drop(rt);
+}

--- a/tokio/CHANGELOG.md
+++ b/tokio/CHANGELOG.md
@@ -1,3 +1,8 @@
+# 1.0.3 (January 28, 2020)
+
+### Fixed
+- io: memory leak during shutdown (#3477).
+
 # 1.0.2 (January 14, 2020)
 
 ### Fixed

--- a/tokio/Cargo.toml
+++ b/tokio/Cargo.toml
@@ -8,12 +8,12 @@ name = "tokio"
 #   - README.md
 # - Update CHANGELOG.md.
 # - Create "v1.0.x" git tag.
-version = "1.0.2"
+version = "1.0.3"
 edition = "2018"
 authors = ["Tokio Contributors <team@tokio.rs>"]
 license = "MIT"
 readme = "README.md"
-documentation = "https://docs.rs/tokio/1.0.2/tokio/"
+documentation = "https://docs.rs/tokio/1.0.3/tokio/"
 repository = "https://github.com/tokio-rs/tokio"
 homepage = "https://tokio.rs"
 description = """

--- a/tokio/src/io/driver/registration.rs
+++ b/tokio/src/io/driver/registration.rs
@@ -207,6 +207,13 @@ impl Registration {
 
 impl Drop for Registration {
     fn drop(&mut self) {
+        // It is possible for a cycle to be created between wakers stored in
+        // `ScheduledIo` instances and `Arc<driver::Inner>`. To break this
+        // cycle, wakers are cleared. This is an imperfect solution as it is
+        // possible to store a `Registration` in a waker. In this case, the
+        // cycle would remain.
+        //
+        // See tokio-rs/tokio#3481 for more details.
         self.shared.clear_wakers();
     }
 }

--- a/tokio/src/io/driver/registration.rs
+++ b/tokio/src/io/driver/registration.rs
@@ -205,6 +205,12 @@ impl Registration {
     }
 }
 
+impl Drop for Registration {
+    fn drop(&mut self) {
+        self.shared.clear_wakers();
+    }
+}
+
 fn gone() -> io::Error {
     io::Error::new(io::ErrorKind::Other, "IO driver has terminated")
 }

--- a/tokio/src/io/driver/scheduled_io.rs
+++ b/tokio/src/io/driver/scheduled_io.rs
@@ -355,6 +355,12 @@ impl ScheduledIo {
         // result isn't important
         let _ = self.set_readiness(None, Tick::Clear(event.tick), |curr| curr - mask_no_closed);
     }
+
+    pub(crate) fn clear_wakers(&self) {
+        let mut waiters = self.waiters.lock();
+        waiters.reader.take();
+        waiters.writer.take();
+    }
 }
 
 impl Drop for ScheduledIo {

--- a/tokio/src/lib.rs
+++ b/tokio/src/lib.rs
@@ -1,4 +1,4 @@
-#![doc(html_root_url = "https://docs.rs/tokio/1.0.2")]
+#![doc(html_root_url = "https://docs.rs/tokio/1.0.3")]
 #![allow(
     clippy::cognitive_complexity,
     clippy::large_enum_variant,

